File: src/main/java/game/util/GameUtil.java
package game.util;

import game.GameServer;
import game.io.Requests.MinionAggroRequest;
import game.io.Requests.UpdateHeroItemPositionRequest;
import game.io.Responses.CombatTextResponse;
import game.logging.Log;
import game.vo.*;
import game.vo.classes.Priest;
import game.vo.classes.Warlock;
import game.vo.classes.Warrior;

import java.text.MessageFormat;
import java.util.*;

import static java.lang.Thread.sleep;

/**
 * Created by Eric on 2017-02-02.
 */
public class GameUtil {
	private static final String TAG = GameUtil.class.getSimpleName();
	public static final String COLOR_DAMAGE = "#FFFFFFFF";
	public static final String COLOR_HEAL = "#FF00FF00";
	public static final String COLOR_CRIT = "#FFFF0000";

	private GameServer gameServer;
	private static int itemsForEachLevel = 5;

	public GameServer getGameServer() {
		return gameServer;
	}

	public void setGameServer(GameServer gameServer) {
		this.gameServer = gameServer;
	}

	public static boolean isWorldType(int type, int worldLevel) {
		if(type == getWorldTypeFromLevel(worldLevel)){
			return true;
		}
		return false;
	}

	public static int getWorldTypeFromLevel(int worldLevel) {
		int worldType = worldLevel % 5;
		if(worldType == 4 || worldType == 0 ){
			worldType = World.CRAWLER;
		}
		return worldType;
	}


	public void startHordeMinionSpawner(GameServer server) {
		Thread thread = new Thread(() -> {
			while (server.getHordeMinionsLeft() > 0) {
				Point point = getRandomEnemySpawnPoint(server);
				if (point != null) {
					Minion minion = server.spawnMinion(point.getPosX(), point.getPosZ(), point.getPosY());
					minion.setDesiredPositionX(0);
					minion.setDesiredPositionZ(0);
				} else {
					Log.i(TAG, "We did not get a spawn point, something is wrong");
				}
				server.setHordeMinionsLeft(server.getHordeMinionsLeft() -1);
				try {
					sleep(5000);
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
		});
		thread.start();
	}

	private Point getRandomEnemySpawnPoint(GameServer server) {
		Point point = null;
		while (point == null) {
			int randomPos = CalculationUtil.getRandomInt(0, (server.getWorld().getSpawnPoints().size() - 1));
			point = server.getWorld().getSpawnPoints().get(randomPos);
			Log.i(TAG, "Trying to find a random spawn point at pos:  " + randomPos + " its of type : " + point.getPointType());
			if (point.getPointType() == Point.SPAWN_POINT) {
				point = null;
			}
		}
		return point;
	}


	public void minionTargetInRange(MinionAggroRequest parsedRequest) {
		if (gameServer != null && parsedRequest.getMinion_id() != null) {
			Minion minion = getMinionById(parsedRequest.getMinion_id(), gameServer.getMinions());
			if (minion != null && parsedRequest.getHero_id() > 0) {
//				Log.i(TAG, "Target is in range for an attack");
				minion.targetInRangeForAttack = true;
			} else {
				Log.i(TAG, "Target is out of range for an attack");
				if (minion != null) {
					minion.targetInRangeForAttack = false;
				}
			}
		}else{
			Log.i(TAG, "Something wrong, either gameserver is null or minion_id is");
			if (gameServer == null) {
				Log.i(TAG, "GameServer is null");
			}
		}
	}

	/**
	 * Goes through the list of heroes and returning the hero that has the lowest hp
	 *
	 * @return hero
	 */
	public Hero getHeroWithLowestHp() {
		Hero targetHero = null;

		if (gameServer != null && gameServer.getHeroes() != null) {
			for (Hero lowestHpHero : gameServer.getHeroes()) {
				if (targetHero == null) {
					targetHero = lowestHpHero;
				} else if (lowestHpHero.getHp() < targetHero.getHp()) {
					targetHero = lowestHpHero;
				}
			}
		}
		return targetHero;
	}

	public static ArrayList<Item> generateLoot(Hero hero) {
		ArrayList<Item> loot = new ArrayList<>();
		ArrayList<Item> items = null;
		if (hero != null) {
			items = DatabaseUtil.getItems(hero.getLevel());
		}

		if(items == null){
			items = new ArrayList<>();
		}
		// Check if there is enough items to roll otherwise create new items
		if(items.size() < itemsForEachLevel){
			Log.i(TAG, "Not enough items to loot through, generate a couple more. " + items.size() + " of " + itemsForEachLevel);
			for(int i = 0; i < (itemsForEachLevel - items.size()) ; i++ ){
				Item item = ItemUtil.generateItem(hero.getLevel(), hero);
				Log.i(TAG, "Generated item cause there was not enough in database.");
				items.add(item);
				DatabaseUtil.addItem(item);
			}
		}

		// Roll if hero got these items
		for (Item item : items) {
			int rate = CalculationUtil.getRandomInt(0,1000);
			Log.i(TAG, "Drop rate : " + rate + " /" + Math.round(item.getDropRate() * 1000));
			if(rate <= (item.getDropRate() * 1000)){
				item.generateInfo(true);
				item.setHeroId(hero.getId());
				DatabaseUtil.addLoot(item);
				loot.add(item);
			}
		}
		return loot;
	}

	public static Hero getHeroByUserId(String user_id, ArrayList<Hero> heroes) {
		for (Hero hero : heroes) {
			if (hero.getUser_id() != null && hero.getUser_id() == Integer.parseInt(user_id)) {
				return hero;
			}
		}
		return null;
	}

	public static Hero getHeroById(Integer heroId, ArrayList<Hero> heroes) {
		for (Hero hero : heroes) {
			if (hero.getId() == heroId) {
				return hero;
			}
		}
		return null;
	}

	public static Minion getMinionById(Integer minionId, ArrayList<Minion> minions) {
		for (Minion minion : minions) {
			if (minion.getId() == minionId) {
				return minion;
			}
		}
		return null;
	}

	public static void removeMinion(Integer minionId, ArrayList<Minion> minions) {
		Iterator<Minion> ut = minions.iterator();
		while (ut.hasNext()) {
			Minion minion = ut.next();
			if (minion.getId() == minionId) {
				ut.remove();
				removeMinion(minion.getId(), minions);
			}
		}
	}

	private void setTalents(Hero hero, ArrayList<Talent> talents) {
		hero.setTalents(talents);
	}


	public void updateItemPosition(String userId, UpdateHeroItemPositionRequest request) {
		Log.i(TAG, "Updated item " + request.getItemId() + " To pos: " + request.getNewPosition());
		DatabaseUtil.updateHeroItem(request.getItemId(), request.getNewPosition(), request.getEquipped());
		getHeroById(request.getHeroId(), getGameServer().getHeroes()).updateStats();
		getGameServer().sendGameStatus();
	}


	public static void minionAggro(MinionAggroRequest parsedRequest, ArrayList<Minion> minions) {
		Minion minion = GameUtil.getMinionById(parsedRequest.getMinion_id(), minions);
		if (minion != null && minion.getHeroIdWithMostThreat() == null) {
			Log.i(TAG, "This minion had no aggro, add towards this hero [" + parsedRequest.getHero_id() + "] Since first to see it");
			minion.addThreat(new Threat(parsedRequest.getHero_id(), Threat.inRangeThreath, 0, 0));
		}else{
			if (minion != null) {
				Log.i(TAG, "Minion had aggro : " + minion.getHeroIdWithMostThreat());
			} else {
				Log.i(TAG, "Did not find minion with id : " + parsedRequest.getMinion_id());
			}
		}
	}

	public static void updateMinionPositions(ArrayList<Minion> updatedMinions, ArrayList<Minion> minions) {
		// TODO: Do this more efficient
		for (Minion minion : updatedMinions) {
			for (Minion gameMinion : minions) {
				if (minion.getId() == gameMinion.getId()) {
					gameMinion.setPositionX(minion.getPositionX());
					gameMinion.setPositionY(minion.getPositionY());
					gameMinion.setPositionZ(minion.getPositionZ());
				}
			}
		}
	}

	public void dealDamageToMinion(Hero hero, Minion minion, float damage) {
		if (minion.takeDamage(damage)) {
			Log.i(TAG, "Found minion to attack : " + minion.getId() + " Minion died!!!");
			getGameServer().minionDied(hero.getId(), minion.getId());
			// Send stop movement to all attacking this minion
			HeroUtil.stopHero(hero.id, getGameServer());
		} else {
			Log.i(TAG, "Found minion to attack : " + minion.getId() + " new hp is: " + minion.getHp());
			minion.addThreat(new Threat(hero.getId(), 0.0f, damage, 0.0f));
		}
	}


	/**
	 * A hero has clicked a portal, check that all have then start new level
	 *
	 * @param heroId
	 */
	public void clickPortal(int heroId, ArrayList<Hero> heroes) {
		Log.i(TAG, "This hero has clicked on the stairs " + heroId);
		Hero hero = GameUtil.getHeroById(heroId, heroes);
		if (hero != null) {
			hero.setStairsPressed();
		}
		boolean startNewGame = true;
		for (Hero heroInList : heroes) {
			if (!heroInList.isStairsPressed()) {
				startNewGame = false;
			}
		}
		// Start new level
		if (startNewGame) {
			getGameServer().gameLevel = getGameServer().gameLevel + 1;
			getGameServer().startLevel();
		} else {
			Log.i(TAG, "Not all heroes has clicked the portal");
		}
	}


	/**
	 * Loops through the different spawn points that's available to start on
	 * @param world
	 */
	public static Vector3 getFreeStartPosition(World world) {
		Collections.shuffle(world.getSpawnPoints(), new Random(System.nanoTime()));
		for (Point point : world.getSpawnPoints()) {
			if (!point.isUsed() && point.getPointType() == Point.SPAWN_POINT) {
				Vector3 spawnPoint = point.getLocation();
				point.setUsed(true);
				Log.i(TAG, "Found spawnpoint to use: " + point.toString());
				return spawnPoint;
			}
		}
		return null;
	}


	/**
	 * When a player join he needs to have a hero, this makes sure his hero join as the correct class etc.
	 *
	 * @param hero
	 */
	public void addHero(Hero hero, ArrayList<Hero> heroes) {
		getGameServer().sendWorldOperation(hero.getId());
		if (hero.isClass(Hero.WARRIOR)) {
			Log.i(TAG, "Added a warrior");
			Warrior warrior = (Warrior) hero;
			warrior.generateHeroInformation();
			heroes.add(warrior);
		} else if (hero.isClass(Hero.PRIEST)) {
			Log.i(TAG, "Added a priest");
			Priest priest = (Priest) hero;
			priest.generateHeroInformation();
			heroes.add(priest);
		} else if (hero.isClass(Hero.WARLOCK)) {
			Log.i(TAG, "Added a warlock");
			Warlock warlock = (Warlock) hero;
			warlock.generateHeroInformation();
			heroes.add(warlock);
		}
		Log.i(TAG, "Hero joined with this user id: " + hero.getUser_id() + " characters in game: " + heroes.size());
		getGameServer().sendGameStatus();
		CommunicationUtil.sendAbilities("" + hero.getUser_id(), getGameServer());
		CommunicationUtil.sendTalents(hero.getUser_id(), getGameServer());
		HeroUtil.getHeroItems(hero, true, getGameServer());

		getGameServer().setGameStarted();
	}

	public void addSpawnPoints(ArrayList<Point> points) {
		boolean pointAdded = false;
		Log.i(TAG, "Got this many spawnpoints : " + points.size());
		for (Point point : points) {
			if (getGameServer().getWorld().getSpawnPoint(point.getLocation()) == null) {
				Log.i(TAG, "Adding point of type : " + point.getPointType() + " position: " + point.getLocation().toString());
				if (point.getPointType() == Point.SPAWN_POINT) {
					pointAdded = true;
					getGameServer().getWorld().addSpawPoint(point);
				} else if (point.getPointType() == Point.ENEMY_POINT) {
					getGameServer().getWorld().addSpawPoint(point);
					if (getGameServer().getWorld().getWorldType() != World.HORDE) {
						getGameServer().spawnMinion(point.getPosX(), point.getPosZ(), point.getPosY());
					}
				}
			}
		}

		if (getGameServer().getWorld().getWorldType() == World.HORDE) {
			startHordeMinionSpawner(getGameServer());
		}

		if (pointAdded) {
			Log.i(TAG, "Sending new spawn location for heroes");
			for (Hero hero : getGameServer().getHeroes()) {
				Vector3 location = GameUtil.getFreeStartPosition(getGameServer().getWorld());
				hero.setPositionX(location.getX());
				hero.setPositionZ(location.getZ());
				hero.setPositionY(location.getY());
				hero.setDesiredPositionY(location.getY());
				hero.setDesiredPositionX(location.getX());
				hero.setDesiredPositionZ(location.getZ());
				Log.i(TAG, "Setting new location for hero " + hero.getId() + " " + hero.getPositionX() + "x" + hero.getPositionZ() + "y" + hero.getPositionY());
			}
			CommunicationUtil.sendTeleportPlayers(getGameServer());
		}
	}

	public void minionDebuffs(ArrayList<Minion> minions, ArrayList<Hero> heroes) {
		for (Minion minion : minions) {
			if (minion.isAlive()) {
				// Take action for debuffs
				if (minion.getDeBuffs().size() > 0) {
					Log.i(TAG, "Debuffs left : " + minion.getDeBuffs().size());
					Iterator<Buff> iterator = minion.getDeBuffs().iterator();
					while (iterator.hasNext()) {
						Buff debuff = iterator.next();
						long tickTimeConv = Long.parseLong(debuff.tickTime);
						if (tickTimeConv > 0 && (System.currentTimeMillis() >= tickTimeConv)) {
							Log.i(TAG, "It was time for minion debuff changed debuff tick time to " + (debuff.tickTime + debuff.duration) + " from " + debuff.tickTime);
							debuff.tickTime = "" + (tickTimeConv + debuff.duration);
							if (debuff.type == Buff.DOT) {
								dealDamageToMinion(GameUtil.getHeroById(debuff.heroId, heroes), minion, debuff.value);

								CommunicationUtil.sendCombatText(new CombatTextResponse(false, minion.getId(), "" + debuff.value, false, COLOR_DAMAGE), getGameServer());

								debuff.ticks--;
								if (debuff.ticks == 0) {
									iterator.remove();
								}
							}
						}
					}
				}
			}
		}
	}

	public void heroBuffs(ArrayList<Minion> minions, ArrayList<Hero> heroes) {
		Log.i(TAG, "Got tick to go through hero buffs");
		for(Hero hero : heroes){
			if(hero.getBuffs().size() > 0){
				Iterator<Buff> iterator = hero.getBuffs().iterator();
				while (iterator.hasNext()) {
					Buff buff = iterator.next();
					Log.i(TAG, "Hero [" + hero.getId() + "] got buffs : " + hero.getBuffs());
					if (buff.type == Buff.HOT) {
						Log.i(TAG, "Healed for " + buff.value);
						hero.heal(new Amount(buff.value));
						CommunicationUtil.sendCombatText(new CombatTextResponse(true, hero.getId(), "" + buff.value, false, COLOR_HEAL), getGameServer());

						buff.ticks--;
						if (buff.ticks == 0) {
							iterator.remove();
						}
					}
				}
			}
		}
	}
}

File: src/main/java/game/util/GameUtil.kt
package game.util

import game.GameServer
import game.io.CommunicationHandler
import game.io.Requests.MinionAggroRequest
import game.io.Requests.UpdateHeroItemPositionRequest
import game.io.Responses.CombatTextResponse
import game.logging.Log
import game.logic.math.CalculationUtil
import game.models.*
import game.models.abilities.Buff
import game.models.enemies.Minion
import game.models.game.Amount
import game.models.game.Point
import game.models.game.Threat
import game.models.game.World
import game.models.heroes.Hero
import game.models.heroes.priest.Priest
import game.models.heroes.warlock.Warlock
import game.models.heroes.warrior.Warrior
import game.models.talents.Talent
import game.models.items.Item
import game.models.vo.Vector3
import java.util.*

/**
 * Created by Eric on 2017-02-02.
 */
class GameUtil {
    var gameServer: GameServer? = null

    private fun startHordeMinionSpawner(server: GameServer) {
        val thread = Thread {
            while (server.hordeMinionsLeft > 0) {
                val point = getRandomEnemySpawnPoint(server)
                val minion = server.spawnMinion(point.posX, point.posZ, point.posY)
                minion.desiredPositionX = 0f
                minion.desiredPositionZ = 0f
                server.hordeMinionsLeft = server.hordeMinionsLeft - 1
                try {
                    Thread.sleep(5000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }
        thread.start()
    }

    private fun getRandomEnemySpawnPoint(server: GameServer): Point {
        var point: Point? = null
        while (point == null) {
            val randomPos = CalculationUtil.getRandomInt(0, server.world!!.spawnPoints.size - 1)
            point = server.world!!.spawnPoints[randomPos]
            Log.i(
                TAG,
                "Trying to find a random spawn point at pos:  " + randomPos + " its of type : " + point.pointType
            )
            if (point.pointType == Point.SPAWN_POINT) {
                point = null
            }
        }
        return point
    }

    fun minionTargetInRange(parsedRequest: MinionAggroRequest) {
        if (gameServer != null && parsedRequest.minion_id != null) {
            val minion = getMinionById(parsedRequest.minion_id, gameServer!!.minions)
            if (minion != null && parsedRequest.hero_id > 0) {
//				Log.i(TAG, "Target is in range for an attack");
                minion.targetInRangeForAttack = true
            } else {
                Log.i(TAG, "Target is out of range for an attack")
                if (minion != null) {
                    minion.targetInRangeForAttack = false
                }
            }
        } else {
            Log.i(TAG, "Something wrong, either gameserver is null or minion_id is")
            if (gameServer == null) {
                Log.i(TAG, "GameServer is null")
            }
        }
    }

    /**
     * Goes through the list of heroes and returning the hero that has the lowest hp
     *
     * @return hero
     */
    val heroWithLowestHp: Hero?
        get() {
            var targetHero: Hero? = null
            if (gameServer != null) {
                for (lowestHpHero in gameServer!!.heroes) {
                    if (targetHero == null) {
                        targetHero = lowestHpHero
                    } else if (lowestHpHero.hp < targetHero.hp) {
                        targetHero = lowestHpHero
                    }
                }
            }
            return targetHero
        }

    private fun setTalents(hero: Hero, talents: ArrayList<Talent>) {
        hero.talents = talents
    }

    fun updateItemPosition(request: UpdateHeroItemPositionRequest) {
        Log.i(TAG, "Updated item " + request.itemId + " To pos: " + request.newPosition)
        DatabaseUtil.updateHeroItem(request.itemId.toLong(), request.newPosition, request.equipped)
        gameServer?.heroes?.let {
            getHeroById(request.heroId, it)
                ?.updateStats()
        }
        gameServer?.sendGameStatus()
    }

    fun dealDamageToMinion(hero: Hero?, minion: Minion, damage: Float) {
        if (minion.takeDamage(damage)) {
            Log.i(TAG, "Found minion to attack : " + minion.id + " Minion died!!!")
            gameServer!!.minionDied(hero!!.getId(), minion.id)
            // Send stop movement to all attacking this minion
            HeroUtil.stopHero(hero.id, gameServer)
        } else {
            Log.i(TAG, "Found minion to attack : " + minion.id + " new hp is: " + minion.hp)
            minion.addThreat(Threat(hero!!.getId(), 0.0f, damage, 0.0f))
        }
    }

    /**
     * A hero has clicked a portal, check that all have then start new level
     *
     * @param heroId
     */
    fun clickPortal(heroId: Int, heroes: ArrayList<Hero>) {
        Log.i(TAG, "This hero has clicked on the stairs $heroId")
        val hero = getHeroById(heroId, heroes)
        hero?.setStairsPressed()
        var startNewGame = true
        for (heroInList in heroes) {
            if (!heroInList.isStairsPressed) {
                startNewGame = false
            }
        }
        // Start new level
        if (startNewGame) {
            gameServer?.let{ server ->
                server.gameLevel = server.gameLevel + 1
                server.startLevel()
            }
        } else {
            Log.i(TAG, "Not all heroes has clicked the portal")
        }
    }

    /**
     * When a player join he needs to have a hero, this makes sure his hero join as the correct class etc.
     *
     * @param hero
     */
    fun addHero(hero: Hero, heroes: ArrayList<Hero?>) {
        gameServer?.sendWorldOperation(hero.getId())
        if (hero.isClass(Hero.WARRIOR)) {
            Log.i(TAG, "Added a warrior")
            val warrior = hero as Warrior
            warrior.generateHeroInformation()
            heroes.add(warrior)
        } else if (hero.isClass(Hero.PRIEST)) {
            Log.i(TAG, "Added a priest")
            val priest = hero as Priest
            priest.generateHeroInformation()
            heroes.add(priest)
        } else if (hero.isClass(Hero.WARLOCK)) {
            Log.i(TAG, "Added a warlock")
            val warlock = hero as Warlock
            warlock.generateHeroInformation()
            heroes.add(warlock)
        }
        Log.i(TAG, "Hero joined with this user id: " + hero.user_id + " characters in game: " + heroes.size)
        gameServer?.sendGameStatus()
        CommunicationHandler.sendAbilities("" + hero.user_id, gameServer)
        CommunicationHandler.sendTalents(hero.user_id, gameServer)
        HeroUtil.getHeroItems(hero, true, gameServer)
        gameServer?.setGameStarted()
    }

    fun addSpawnPoints(points: ArrayList<Point>) {
        var pointAdded = false
        Log.i(TAG, "Got this many spawnpoints : " + points.size)
        for (point in points) {
            if (gameServer?.world?.getSpawnPoint(point.location) == null) {
                Log.i(TAG, "Adding point of type : " + point.pointType + " position: " + point.location.toString())
                if (point.pointType == Point.SPAWN_POINT) {
                    pointAdded = true
                    gameServer?.world?.addSpawPoint(point)
                } else if (point.pointType == Point.ENEMY_POINT) {
                    gameServer?.world?.addSpawPoint(point)
                    if (gameServer?.world?.worldType != World.HORDE) {
                        gameServer?.spawnMinion(point.posX, point.posZ, point.posY)
                    }
                }
            }
        }
        if (gameServer?.world?.worldType == World.HORDE) {
            gameServer?.let{
                startHordeMinionSpawner(it)
            }
        }
        if (pointAdded) {
            Log.i(TAG, "Sending new spawn location for heroes")
            gameServer?.let{ gameServer ->
                gameServer.heroes.let{ heroes ->
                    for (hero in heroes) {
                        gameServer.world?.let{ world ->
                            val location = getFreeStartPosition(world)
                            hero.positionX = location!!.x
                            hero.positionZ = location.z
                            hero.positionY = location.y
                            hero.desiredPositionY = location.y
                            hero.desiredPositionX = location.x
                            hero.desiredPositionZ = location.z
                            Log.i(
                                TAG,
                                "Setting new location for hero " + hero.getId() + " " + hero.positionX + "x" + hero.positionZ + "y" + hero.positionY
                            )
                        }
                    }
                }
                CommunicationHandler.sendTeleportPlayers(gameServer)
            }
        }
    }

    fun handleMinionDebuffs(minions: ArrayList<Minion>, heroes: ArrayList<Hero>) {
        for (minion in minions.filter { it.isAlive }) {
            if (minion.deBuffs.isNotEmpty()) {
                val iterator = minion.deBuffs.iterator()
                while (iterator.hasNext()) {
                    val debuff = iterator.next()
                    val tickTimeMillis = debuff.tickTime.toLongOrNull() ?: continue
                    if (System.currentTimeMillis() >= tickTimeMillis) {
                        debuff.tickTime = (tickTimeMillis + debuff.duration).toString()
                        handleDebuffEffect(debuff, minion, heroes)
                        debuff.ticks--
                        if (debuff.ticks == 0) {
                            iterator.remove()
                        }
                    }
                }
                Log.i(TAG, "Minion has ${minion.deBuffs.size} debuffs remaining")
            }
        }
    }

    private fun handleDebuffEffect(debuff: Buff, minion: Minion, heroes: ArrayList<Hero>) {
        when (debuff.type) {
            Buff.DOT -> {
                val hero = getHeroById(debuff.heroId, heroes) ?: return
                dealDamageToMinion(hero, minion, debuff.value.toFloat())
                gameServer?.let { server ->
                    CommunicationHandler.sendCombatText(
                        CombatTextResponse(false, minion.id, debuff.value.toString(), false, COLOR_DAMAGE),
                        server
                    )
                }
            }
            else -> {
                // Do nothing, since the debuff type is not recognized
            }
        }
    }

    fun heroBuffs(heroes: ArrayList<Hero>) {
        Log.i(TAG, "Got tick to go through hero buffs")
        for (hero in heroes) {
            if (hero.buffs.size > 0) {
                val iterator = hero.buffs.iterator()
                while (iterator.hasNext()) {
                    val buff = iterator.next()
                    Log.i(TAG, "Hero [" + hero.getId() + "] got buffs : " + hero.buffs)
                    if (buff.type == Buff.HOT) {
                        Log.i(TAG, "Healed for " + buff.value)
                        hero.heal(Amount(buff.value.toFloat()))
                        gameServer?.let { server ->
                            CommunicationHandler.sendCombatText(
                                CombatTextResponse(
                                    true,
                                    hero.getId(),
                                    "" + buff.value,
                                    false,
                                    COLOR_HEAL
                                ), server
                            )
                        }
                        buff.ticks--
                        if (buff.ticks == 0) {
                            iterator.remove()
                        }
                    }
                }
            }
        }
    }

    companion object {
        private val TAG = GameUtil::class.java.simpleName
        const val COLOR_DAMAGE = "#FFFFFFFF"
        const val COLOR_HEAL = "#FF00FF00"
        const val COLOR_CRIT = "#FFFF0000"
        private const val itemsForEachLevel = 5
        @JvmStatic
		fun isWorldType(type: Int, worldLevel: Int): Boolean {
            return type == getWorldTypeFromLevel(worldLevel)
        }

        @JvmStatic
		fun getWorldTypeFromLevel(worldLevel: Int): Int {
            var worldType = worldLevel % 5
            if (worldType == 4 || worldType == 0) {
                worldType = World.CRAWLER
            }
            return worldType
        }

        @JvmStatic
		fun generateLoot(hero: Hero?): ArrayList<Item> {
            val loot = ArrayList<Item>()
            var items: ArrayList<Item>? = null
            if (hero != null) {
                items = DatabaseUtil.getItems(hero.level)
            }
            if (items == null) {
                items = ArrayList()
            }
            // Check if there is enough items to roll otherwise create new items
            if (items.size < itemsForEachLevel) {
                Log.i(
                    TAG,
                    "Not enough items to loot through, generate a couple more. " + items.size + " of " + itemsForEachLevel
                )
                for (i in 0 until itemsForEachLevel - items.size) {
                    val item = ItemUtil.generateItem(hero!!.level, hero)
                    Log.i(TAG, "Generated item cause there was not enough in database.")
                    items.add(item)
                    DatabaseUtil.addItem(item)
                }
            }

            // Roll if hero got these items
            for (item in items) {
                val rate = CalculationUtil.getRandomInt(0, 1000)
                Log.i(TAG, "Drop rate : " + rate + " /" + Math.round(item.dropRate * 1000))
                if (rate <= item.dropRate * 1000) {
                    item.generateInfo(true)
                    item.heroId = hero?.getId()?.toLong() ?: 0
                    DatabaseUtil.addLoot(item)
                    loot.add(item)
                }
            }
            return loot
        }

        @JvmStatic
		fun getHeroByUserId(user_id: String, heroes: ArrayList<Hero>): Hero? {
            for (hero in heroes) {
                if (hero.user_id != null && hero.user_id == user_id.toInt()) {
                    return hero
                }
            }
            return null
        }

        @JvmStatic
		fun getHeroById(heroId: Int, heroes: ArrayList<Hero>): Hero? {
            for (hero in heroes) {
                if (hero.getId() == heroId) {
                    return hero
                }
            }
            return null
        }

        @JvmStatic
		fun getMinionById(minionId: Int, minions: ArrayList<Minion>): Minion? {
            for (minion in minions) {
                if (minion.id == minionId) {
                    return minion
                }
            }
            return null
        }

        fun removeMinion(minionId: Int, minions: ArrayList<Minion>) {
            val ut = minions.iterator()
            while (ut.hasNext()) {
                val minion = ut.next()
                if (minion.id == minionId) {
                    ut.remove()
                    removeMinion(minion.id, minions)
                }
            }
        }

        @JvmStatic
		fun minionAggro(parsedRequest: MinionAggroRequest, minions: ArrayList<Minion>) {
            val minion = getMinionById(parsedRequest.minion_id, minions)
            if (minion != null && minion.heroIdWithMostThreat == null) {
                Log.i(
                    TAG,
                    "This minion had no aggro, add towards this hero [" + parsedRequest.hero_id + "] Since first to see it"
                )
                minion.addThreat(
                    Threat(
                        parsedRequest.hero_id,
                        Threat.inRangeThreath,
                        0f,
                        0f
                    )
                )
            } else {
                if (minion != null) {
                    Log.i(TAG, "Minion had aggro : " + minion.heroIdWithMostThreat)
                } else {
                    Log.i(TAG, "Did not find minion with id : " + parsedRequest.minion_id)
                }
            }
        }

        @JvmStatic
		fun updateMinionPositions(updatedMinions: ArrayList<Minion>, minions: ArrayList<Minion>) {
            // TODO: Do this more efficient
            for (minion in updatedMinions) {
                for (gameMinion in minions) {
                    if (minion.id == gameMinion.id) {
                        gameMinion.positionX = minion.positionX
                        gameMinion.positionY = minion.positionY
                        gameMinion.positionZ = minion.positionZ
                    }
                }
            }
        }

        /**
         * Loops through the different spawn points that's available to start on
         * @param world
         */
        fun getFreeStartPosition(world: World): Vector3? {
            world.spawnPoints?.shuffle(Random(System.nanoTime()))
            for (point in world.spawnPoints) {
                if (!point.isUsed && point.pointType == Point.SPAWN_POINT) {
                    val spawnPoint = point.location
                    point.isUsed = true
                    Log.i(TAG, "Found spawnpoint to use: $point")
                    return spawnPoint
                }
            }
            return null
        }
    }
}